### PR TITLE
unify challenge provider logic

### PR DESCRIPTION
**Pull request by schurzi**
restructure the code, so it becomes easier to include additional DNS
challenge providers

### CODE REVIEW

**Comment by rndmh3ro** [comment]
I like this! More order is always better!

I think we could simplify it and make it more robust:

```
- name: check if a challenge provider is used
  assert:
    that:
      - letsencrypt_do_dns_challenge | bool or letsencrypt_do_http_challenge | bool
    fail_msg: "You need to set letsencrypt_do_dns_challenge'' or 'letsencrypt_do_http_challenge'."

- name: check if a challenge provider is used
  assert:
    that:
      - letsencrypt_dns_provider is defined
    fail_msg: "You need to set a dns challenge provider if you use the dns challenge."
  when:
    - letsencrypt_do_dns_challenge is defined
    - letsencrypt_do_dns_challenge | bool

- name: check if a challenge provider is used
  assert:
    that:
      - letsencrypt_http_provider is defined
    fail_msg: "You need to set a http challenge provider if you use the http challenge."
  when:
    - letsencrypt_do_http_challenge is defined
    - letsencrypt_do_http_challenge | bool
```

This checks if one of the challenges is set and if an approriate provider is used.

**Comment by schurzi** [comment]
I adapted your approach a little bit. Hope that is ok. And I found, we did not have a split our http provider logic, so I created this.

**Comment by rndmh3ro** [comment]
> I adapted your approach a little bit. Hope that is ok. 

Sure, it was just a general concept. 

> And I found, we did not have a split our http provider logic, so I created this. 

We will soon have two http providers.

File: roles/letsencrypt/tasks/http-challenge.yml [hunk 1]
```diff
@@ -28,73 +20,8 @@
...
-    - name: remove challenge file for SAN domain from fs
-      file:
-        dest: "acme-challenge.{{ item }}"
-        state: absent
-      loop: "{{ domain.subject_alt_name }}"
-      when:
-        - domain.subject_alt_name is defined
-        # only runs if the challenge is run the first time, because then there is challenge_data
-        - challenge['challenge_data'][item] is defined
+- name: "do dns challenge with provider {{ letsencrypt_http_provider }}"
```
Comment (rndmh3ro) [inline roles/letsencrypt/tasks/http-challenge.yml:23]: Http

**Review by rndmh3ro — APPROVED** [review]

**Pull request merged**